### ISSUE TITLE
Kubelet logs start with unstructured logs then transition to structured logs

### ISSUE THREAD

**Issue by rphillips** [kubernetes/kubernetes#100152]
#### What happened:
When starting the Kubelet with ```KUBELET_FLAGS=--logging-format=json hack/local-up-cluster.sh```, the Kubelet begins with unstructured logging and then transitions to structured logging:

```
  Flag --chaos-chance has been deprecated, will be removed in a future version.
  Flag --cloud-provider has been deprecated, will be removed in 1.23, in favor of removing cloud provider code from Kubelet.
  Flag --cloud-config has been deprecated, will be removed in 1.23, in favor of removing cloud provider code from Kubelet.
  Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ f
  I0311 16:53:18.330740  199454 flags.go:59] FLAG: --add-dir-header="false"
  I0311 16:53:18.330802  199454 flags.go:59] FLAG: --address="0.0.0.0"
  I0311 16:53:18.330808  199454 flags.go:59] FLAG: --allowed-unsafe-sysctls="[]"
  I0311 16:53:18.330817  199454 flags.go:59] FLAG: --alsologtostderr="false"
  I0311 16:53:18.330820  199454 flags.go:59] FLAG: --anonymous-auth="true"
  I0311 16:53:18.330825  199454 flags.go:59] FLAG: --application-metrics-count-limit="100"
  I0311 16:53:18.330829  199454 flags.go:59] FLAG: --authentication-token-webhook="false"
  I0311 16:53:18.330832  199454 flags.go:59] FLAG: --authentication-token-webhook-cache-ttl="2m0s"
  I0311 16:53:18.330836  199454 flags.go:59] FLAG: --authorization-mode="AlwaysAllow"
  I0311 16:53:18.330841  199454 flags.go:59] FLAG: --authorization-webhook-cache-authorized-ttl="5m0s"
  I0311 16:53:18.330844  199454 flags.go:59] FLAG: --authorization-webhook-cache-unauthorized-ttl="30s"
  I0311 16:53:18.330848  199454 flags.go:59] FLAG: --azure-container-registry-config=""
  I0311 16:53:18.330851  199454 flags.go:59] FLAG: --boot-id-file="/proc/sys/kernel/random/boot_id"
  I0311 16:53:18.330855  199454 flags.go:59] FLAG: --bootstrap-kubeconfig="/var/run/kubernetes/kubelet.kubeconfig"
  I0311 16:53:18.330859  199454 flags.go:59] FLAG: --cert-dir="/var/lib/kubelet/pki"
  I0311 16:53:18.330862  199454 flags.go:59] FLAG: --cgroup-driver="cgroupfs"
  I0311 16:53:18.330865  199454 flags.go:59] FLAG: --cgroup-root=""
  I0311 16:53:18.330868  199454 flags.go:59] FLAG: --cgroups-per-qos="true"
  I0311 16:53:18.331499  199454 flags.go:59] FLAG: --volume-stats-agg-period="1m0s"
  I0311 16:53:18.331564  199454 feature_gate.go:243] feature gates: &{map[]}
  Flag --chaos-chance has been deprecated, will be removed in a future version.
  Flag --cloud-provider has been deprecated, will be removed in 1.23, in favor of removing cloud provider code from Kubelet.
  Flag --cloud-config has been deprecated, will be removed in 1.23, in favor of removing cloud provider code from Kubelet.
  Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ f
  I0311 16:53:18.334042  199454 feature_gate.go:243] feature gates: &{map[APIServerIdentity:false AllAlpha:false AnyVolumeDataSource:false BalanceAttachedNodeVolumes:false CSIServiceAccountToken:false CSIVolumeHeal
  I0311 16:53:18.334176  199454 feature_gate.go:243] feature gates: &{map[APIServerIdentity:false AllAlpha:false AnyVolumeDataSource:false BalanceAttachedNodeVolumes:false CSIServiceAccountToken:false CSIVolumeHeal
  I0311 16:53:18.855837  199454 mount_linux.go:206] Detected OS with systemd
  {"ts":1615503198855.9631,"msg":"Version: v1.21.0-beta.1.201+82a2a30bc92c5e-dirty\n","v":0}
```

#### What you expected to happen:
I expect the Kubelet to have structured logging through the startup lifecycle.

#### How to reproduce it (as minimally and precisely as possible):
 ```KUBELET_FLAGS=--logging-format=json hack/local-up-cluster.sh```

#### Environment:
- Kubernetes version (use `kubectl version`): master [2eb32e930a30ec66d45ee57795adb06c8d554f18]
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by ehashman** [comment]
Is it possible that these components are not yet migrated? I don't think flags.go or feature_gates.go are on structured logging yet.

**Comment by mengjiao-liu** [comment]
It's true tha and  flags.go and feature_gates.go is not yet log structured, At least I don't see it in the PR that exists

**Comment by serathius** [comment]
This bug is related to JSON implementation, not structured logging. Configuration of JSON format happens late in component bootstrap process.

**Comment by mengjiao-liu** [comment]
> Configuration of JSON format happens late in component bootstrap process.
If so, is it related to this PR #99693 ? 
This PR changes the JSON format registration from component-base/logs to the component-base/config @serathius

**Comment by serathius** [comment]
@mengjiao-liu No, your PR is not related. As a configuration of JSON I meant the moment the configuration is applied and klog switches to use different format. This is done exactly when particular component calls `logOption.Apply()`

**Comment by mengjiao-liu** [comment]
Thanks, I get it

**Comment by voigt** [comment]
Hey there (@rphillips, @serathius, @pacoxu), Bug-Triage here 👋 ,
Given the fact that the code freeze for this release already started, I´d like to suggest moving this issue to the next milestone. wdyt?

**Comment by voigt** [comment]
Due to v1.22 code freeze, I'm moving this issue to the next release.

/milestone v1.23

**Comment by nannancy** [comment]
Hey there (@rphillips, @serathius, @pacoxu), Bug-Triage here 👋 , Just want to make sure this issue is still on track for the 1.23?

**Comment by serathius** [comment]
Hey @pohly, what's the current status of init logs after your recent improvements? Is there anything left to do here?

**Comment by pohly** [comment]
No, this is not fixed. `k8s.io/component-base/cli/flag.PrintFlags` is still being called before `k8s.io/component-base/logs.(*Options).apply`.

The behavior is a bit different, though: while previously it used the wrong default log format, now it uses the wrong default log level and therefore those `FLAG` messages are not printed.

I'd also like to clarify one thing: the "Flag ... has been deprecated, will be removed in a future version" messages are printed by pflag to stderr. Those are not normal log messages and no matter what we do with logging, they will always be plain text.

**Comment by pohly** [comment]
The fix is easy. Whether there's anyone willing to approve a PR is a different question...

**Comment by pohly** [comment]
We can still fix this in 1.23. PR is here: https://github.com/kubernetes/kubernetes/pull/106520
